### ISSUE TITLE
Combined by pipe xpath query fail to match violation node

### ISSUE THREAD

**Issue by timurt** [checkstyle/checkstyle#5868]
The order of xpath queries in the combined query affects the returned result, nodes those satisfy query can be missed.

Steps to reproduce:

**Java class**
```
/var/tmp $ cat InputTest.java
public class InputTest {
    public void test(int a, int b) {
        if (a > b) {

        } else {

        }
    }

    public void foo2() {
        String programmingLanguage = "Java";
        switch (programmingLanguage) {
            case "Java":
            case "C#":
            case "C++":
                String.CASE_INSENSITIVE_ORDER.equals(programmingLanguage + " is");
                break;
            case "C":
                String.CASE_INSENSITIVE_ORDER.equals(programmingLanguage + " is ");
                break;
            default:
                String.CASE_INSENSITIVE_ORDER.equals(programmingLanguage + " is ");
                break;
        }
    }
}
```

**Config file**
```
/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
      <module name="CyclomaticComplexity">
    	 <property name="max" value="0"/>
      </module>
    </module>
</module>
```

**Executing checkstyle**
```
/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -c config.xml InputTest.java
Starting audit...
[ERROR] /var/tmp/InputTest.java:2:5: Cyclomatic Complexity is 2 (max allowed is 0). [CyclomaticComplexity]
[ERROR] /var/tmp/InputTest.java:10:5: Cyclomatic Complexity is 5 (max allowed is 0). [CyclomaticComplexity]
Audit done.
Checkstyle ends with 2 errors.
```

**Generating xpath queries**
For each violation message we generate xpath queries using line number and column number of the violation

```
/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -s 2:5 InputTest.java
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='test']
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='test']/MODIFIERS
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='test']/MODIFIERS/LITERAL_PUBLIC

/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -s 10:5 InputTest.java
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='foo2']
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='foo2']/MODIFIERS
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='foo2']/MODIFIERS/LITERAL_PUBLIC
```

**Updated config file**
`SuppressionXpathFilter` module was added
```
/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
      <module name="CyclomaticComplexity">
    	 <property name="max" value="0"/>
      </module>
      <module name="SuppressionXpathFilter">
	 <property name="file" value="suppressions_xpath.xml"/>
      </module>
    </module>
</module>
```

**Suppressions File**
Let's combine all xpath queries using `|` delimeter and execute checkstyle
```
/var/tmp $ cat suppressions_xpath.xml
<?xml version="1.0"?>

<!DOCTYPE suppressions PUBLIC
    "-//Puppy Crawl//DTD Suppressions Xpath Experimental 1.2//EN"
    "http://checkstyle.sourceforge.net/dtds/suppressions_1_2_xpath_experimental.dtd">

<suppressions>
    <suppress-xpath checks="CyclomaticComplexity"
        query="/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='test'] | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='test']/MODIFIERS | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='test']/MODIFIERS/LITERAL_PUBLIC | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='foo2'] | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='foo2']/MODIFIERS | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='foo2']/MODIFIERS/LITERAL_PUBLIC"/>
</suppressions>
```
**Executing checkstyle**
Two errors should have been suppressed, but for some reason, the first error was not suppressed
```
/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -c config.xml InputTest.java
Starting audit...
[ERROR] /Users/timurtibeyev/Documents/test/CyclomaticComplexityCheck/InputTest.java:2:5: Cyclomatic Complexity is 2 (max allowed is 0). [CyclomaticComplexity]
Audit done.
Checkstyle ends with 1 errors.
```

**Suppressions File**
Let's rearrange queries, move fourth query to second place and test again
```
/var/tmp $ cat suppressions_xpath.xml
<?xml version="1.0"?>

<!DOCTYPE suppressions PUBLIC
    "-//Puppy Crawl//DTD Suppressions Xpath Experimental 1.2//EN"
    "http://checkstyle.sourceforge.net/dtds/suppressions_1_2_xpath_experimental.dtd">

<suppressions>
    <suppress-xpath checks="CyclomaticComplexity"
        query="/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='test'] | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='foo2'] | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='test']/MODIFIERS | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='test']/MODIFIERS/LITERAL_PUBLIC | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='foo2']/MODIFIERS | 
/CLASS_DEF[@text='InputTest']/OBJBLOCK/METHOD_DEF[@text='foo2']/MODIFIERS/LITERAL_PUBLIC"/>
</suppressions>
```

**Executing checkstyle**
No violation occurred

```
/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -c config.xml InputTest.java
Starting audit...
Audit done.
```

-------

<a href="https://www.bountysource.com/issues/58970831-combined-by-pipe-xpath-query-fail-to-match-violation-node">
<img src="https://api.bountysource.com/badge/issue?issue_id=58970831">
</a>

**Comment by romani** [comment]
>move fourth query to second place and test again

Did you mean third to first ?

>Missed elements in the combined xpath query

What element is missed ?  Please update ticket title  or explain what is missed in description.

**Comment by timurt** [comment]
@romani 

> Did you mean third to first ?

No, it is correct

> What element is missed ? Please update ticket title or explain what is missed in description.

Edited title and description. Maybe title I wrote is not self descriptive, the problem is for some complex queries actual result differs from expected

**Comment by HuGanghui** [comment]
@timurt now, when I try to reproduce your problem mentioned in https://github.com/checkstyle/checkstyle/issues/5868#issue-327314517, I can not get the result that you got. now I will get all Two errors should have been suppressed, but for some reason, both two error were not suppressed. 

But if we use the 
```
java $RUN_LOCALE -jar checkstyle-8.34-SNAPSHOT-all.jar -c config.xml -g InputTest.java
```
to get the auto generated suppressions.xml

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suppressions PUBLIC
        "-//Checkstyle//DTD SuppressionXpathFilter Experimental Configuration 1.2//EN"
        "https://checkstyle.org/dtds/suppressions_1_2_xpath_experimental.dtd">
<suppressions>
    <suppress-xpath
            files="InputTest.java"
            checks="CyclomaticComplexityCheck"
            query="/CLASS_DEF[./IDENT[@text='InputTest']]/OBJBLOCK/METHOD_DEF[./IDENT[@text='test']]"/>
    <suppress-xpath
            files="InputTest.java"
            checks="CyclomaticComplexityCheck"
            query="/CLASS_DEF[./IDENT[@text='InputTest']]/OBJBLOCK/METHOD_DEF[./IDENT[@text='foo2']]"/>
</suppressions>
```

then Executing checkstyle, it works. 

```
Starting audit...
Audit done.
```

so do we still need to solve this issue? because we can recommend users to use auto generated suppressions.xml

**Comment by timurt** [comment]
@HuGanghui 

> I will get all Two errors should have been suppressed, but for some reason, both two error were not suppressed

This is because syntax of the xpath queries was changed. Initially `@text` attribute was part of `CLASS_DEF` and `METHOD_DEF` nodes, but now it is part of `IDENT` node only

> to get the auto generated suppressions.xml

Yes, first of all there was no auto generation when issue was created. 

Secondly, issue is about ordering of subqueries in complex xpath query. Ordering of subqueries pipelined with `or` affected results, which is not correct

> so do we still need to solve this issue? because we can recommend users to use auto generated suppressions.xml

I tried to reproduce described tests but it I got correct results. I am not 100% sure, but I think issue was resolved when we added support for different xpath axis [here](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java#L207)

@romani I think issue can be closed
